packages/core-bridge/src/worker.rs

```diff
@@ -18,439 +37,566 @@ use temporal_sdk_core::{
 
     Ok(cx.undefined())
 }
 
-pub fn push_history(mut cx: FunctionContext) -> JsResult<JsUndefined> {
-    let pusher = cx.argument::<JsBox<HistoryForReplayTunnel>>(0)?;
-    let workflow_id = cx.argument::<JsString>(1)?;
-    let history_binary = cx.argument::<JsArrayBuffer>(2)?;
-    let callback = cx.argument::<JsFunction>(3)?;
-    let data = history_binary.as_slice(&cx);
-    match History::decode_length_delimited(data) {
-        Ok(hist) => {
-            let workflow_id = workflow_id.value(&mut cx);
-            if let Err(e) = pusher.get_chan().map(|chan| {
-                pusher
-                    .runtime
-                    .sender
-                    .send(RuntimeRequest::PushReplayHistory {
-                        tx: chan,
-                        pushme: HistoryForReplay::new(hist, workflow_id),
-                        callback: callback.root(&mut cx),
-                    })
-            }) {
-                callback_with_unexpected_error(&mut cx, callback, e)?;
-            }
-            Ok(cx.undefined())
+pub fn worker_finalize_shutdown(mut cx: FunctionContext) -> JsResult<JsPromise> {
+    let worker: Handle<BoxedWorker> = cx.argument(0)?;
+
+    match worker.take() {
+        Some(worker_handle) => {
+            let worker = worker_handle
+                .core_worker
+                .try_borrow_mut()
+                .map_err(|_| "Worker still in use")
+                .and_then(|mut val| {
+                    Arc::try_unwrap(val.take().unwrap()).map_err(|_| "Expected 1 reference")
+                })
+                .expect("Some error");
+
+            worker_handle.runtime_handle.future_to_promise(
+                &mut cx,
+                async move {
+                    worker.finalize_shutdown().await;
+                    Ok(())
+                },
+                move |cx, result: anyhow::Result<()>| match result {
+                    Ok(client_handle) => Ok(cx.boxed(RefCell::new(Some(client_handle)))),
+                    Err(err) => cx.throw_unexpected_error(format!(
+                        "Failed to complete Worker shutdown: {:?}",
+                        err
+                    ))?,
+                },
+            )
         }
-        Err(e) => cx.throw_error(format!("Error decoding history: {:?}", e)),
+        None => cx.throw_illegal_state_error("Worker has already been shutdown"),
     }
 }
 
-pub fn close_history_stream(mut cx: FunctionContext) -> JsResult<JsUndefined> {
-    let pusher = cx.argument::<JsBox<HistoryForReplayTunnel>>(0)?;
-    pusher.shutdown();
-    Ok(cx.undefined())
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub type BoxedHistoryForReplayTunnel = JsBox<RefCell<Option<HistoryForReplayTunnelHandle>>>;
+
+pub(crate) struct HistoryForReplayTunnelHandle {
+    pub(crate) runtime_handle: Arc<RuntimeHandle>, // FIXME: Should we inline rather than Arc?
+    sender: Cell<Option<Sender<HistoryForReplay>>>,
```

Comment:
I don't think you need a cell here, seemingly you can rely on the whole struct getting dropped?